### DISCUSSION TITLE
Implementing DSP capabilities

### DISCUSSION THREAD

**Discussion by AydySmith** [pschatzmann/ESP32-A2DP#215] [Q&A] [answered]
I've really enjoyed using this library, I have managed to connect a device and play audio successfully, but I would also like to filter the audio before it is output. Is it possible, and how would I go about grabbing the data (every sample per frame) and manipulating it before it is output.

**Comment by pschatzmann** [comment, marked as answer]
It is documented in the Readme (Accessing the Sink Data Stream with Callbacks) and there is an example ([bt_music_receiver_datacallback](https://github.com/pschatzmann/ESP32-A2DP/blob/main/examples/bt_music_receiver_datacallback)). Just cast the data to a (int16_t*). The even samples are for one channel the odd samples are for the other channel.

I would recommend that you use my [Arduino Audio Tools](https://github.com/pschatzmann/arduino-audio-tools) library for the for the DSP processing. There are many examples in that library e.g. [basic-a2dp-i2s/](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino)  just search for a2dp

**Reply by AydySmith** [reply]
Thanks so much, I did go through the documentation, I just didn't fully understand it. I have a lot more learning to do.
